vcd/catalogitem.go

```diff
@@ -55,18 +56,30 @@ func findCatalogItem(d *schema.ResourceData, vcdClient *VCDClient) (*govcd.Catal
 
 	catalog, err := adminOrg.GetCatalogByName(d.Get("catalog").(string), false)
 	if err != nil {
-		log.Printf("[DEBUG] Unable to find catalog. Removing from tfstate")
-		d.SetId("")
+		log.Printf("[DEBUG] Unable to find catalog.")
 		return nil, fmt.Errorf("unable to find catalog: %s", err)
 	}
 
-	catalogItem, err := catalog.GetCatalogItemByName(d.Get("name").(string), false)
+	identifier := d.Id()
+
+	// check if identifier still is in deprecated style `catalogName:mediaName`
+	if identifier == "" || strings.Contains(identifier, ":") {
+		identifier = d.Get("name").(string)
+	}
+
+	catalogItem, err := catalog.GetCatalogItemByNameOrId(identifier, false)
 	if err != nil {
 		log.Printf("[DEBUG] Unable to find catalog item. Removing from tfstate")
-		return nil, fmt.Errorf("unable to find catalog item: %s", err)
+		d.SetId("")
+		return nil, nil
+	}
+
+	entityId, err := govcd.GetBareEntityUuid(catalogItem.CatalogItem.ID)
```

Comment:
Why we can't use the simple catalogItem.ID instead of doing this transformation?

Reply:
Cause anyways we will need todo transformation when we use or refernce ID and you have to know, that you need to parse it. So this chage allow to save "384be749-2d58-41e6-b4a0-0ce407748cea" and not "urn:vcloud:media:84be749-2d58-41e6-b4a0-0ce407748cea"

Reply:
It seems to me that we are adding complexity to terraform-provider-vcd instead of dealing with it in go-vcloud-director.

I would expect to have methods that retrieve a media item with all the elements that are needed here, and then we simply use it.
One thing I don't understand is why sometimes we look for MediaItem through a Query and sometimes we use GetCatalogItem.

Reply:
This I think is wider discussion regarding what type of ID we save and fetch from state file. 
Regarding MediaItem and CatalogItem as Media -> you can upload Media to catalog and to vdc. Also mediaitem resides as catalogitem in catalog. Its a mess already in vCD.

Reply:
> This I think is wider discussion regarding what type of ID we save and fetch from state file.

What's the question specifically?

> you can upload Media to catalog and to vdc

Please elaborate on how to upload to VDC only?

Reply:
Question what type of Id save in state file:
urn:vcloud:media:84be749-2d58-41e6-b4a0-0ce407748cea
vs
384be749-2d58-41e6-b4a0-0ce407748cea
----
Seems we have to save in extended ID style and parse UUID when we need, as NSX API requests extended ID style.

Reply:
After analysis - API allows to upload to vCD or catalog. Our current govcd supports only to catalog way.

Reply:
You mean VDC? Interesting! Would you share a screen shot how to do it through UI?

Reply:
> Seems we have to save in extended ID style and parse UUID when we need, as NSX API requests extended ID style.

And what ID styles other resources use?

Reply:
Atleast I need short one cause usually we construct api request URL with short ID.

Reply:
> You mean VDC? Interesting! Would you share a screen shot how to do it through UI?

From UI perspective - in flash I don't have working plugin - so haven't seen myself. In HTML5 UI - found only for Tenant upload capabilities, but both media and OVF requires provide catalog.

Reply:
Oh, so it's API-only capability?

Reply:
I know it's in API and in UI I can't confirm due reasons of flash current compatibility and in new HTML5 I didn't found.

Reply:
@dataclouder @Didainius we discussed the following three points in-depth with Vaidotas and converged to the following conclusions:

1. Use the `urn:...` long ID format, because we want to be a thin layer above vCD API where possible. Granted, this adds additional `GetBareEntityUuid` calls inside the code.
2. Continue using `Query...` functions where it makes sense. `Query...` functions are by agreement those, which call upon the `/query` API endpoint. Most of them return array, but some may return a filtered single record. We already have both return types in code, but it's no issue. In general, we had an agreement that we have `Get` and `Query` function classes some time ago. In this case it's easier to use /query, because it's only a single call to get the media, as opposed to getting it through the catalog.
3. Review which media functions may need to be deprecated. There's one `Find...` function that certainly has a bug - it errors out if there are two media items with same name in two catalogs. We want to move away from it.

Next step is for Vaidotas to refactor where appropriate given the above.

Reply:
This decision goes against our own Coding guidelines. (https://github.com/vmware/go-vcloud-director/blob/master/CODING_GUIDELINES.md)

Reply:
I already added all Get* functions by code guidelines, Query is additional one to get more data.

Reply:
> This decision goes against our own Coding guidelines
It's not good that coding guidelines do not describe Query parameters. We had an agreement that for /query endpoints we keep `Query...` functions. I don't see how `Get...` functions can replace `Query...` functions, especially those which return an array. Please amend the coding guidelines or let's discuss.

Reply:
There is no conflict here.
We agreed that the `Get*` functions are used to retrieve single entities, while `Query*` functions retrieve arrays.
If the function retrieve a single entity, let's call it `GetEntityBySomething` and use it consistently.

Reply:
Updated by new govcd changes